### PR TITLE
Avoid trying to remove empty lists of transactions

### PR DESCRIPTION
**Pull request by jasagredo**
# Description

By increasing the opcert twice on a testing cluster, we found out that this code path could be executed with an empty list of transactions leading to an exception. This change make sure we only try to remove transactions if there are any in the block, and in fact, only of the error of the block was due to the body.

### CODE REVIEW

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs [hunk 1]
```diff
@@ -451,15 +451,25 @@ forkBlockForging IS{..} blockForging =
...
+            -- will remove all the transactions in that block from the mempool
+            -- as a defensive programming measure. Otherwise we'll run the risk
+            -- of forging the same invalid block again. This means that we'll
+            -- throw away some good transactions in the process.
+            case reason of
+              ChainDB.ValidationError (ExtValidationErrorLedger _) ->
+                maybe
+                  (pure ())
+                  (lift . removeTxs mempool)
+                  $ NE.nonEmpty (map (txId . txForgetValidated) txs)
```
Comment (amesgen) [inline ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs:471]:
```suggestion
                whenJust
                  (NE.nonEmpty (map (txId . txForgetValidated) txs))
                  (lift . removeTxs mempool)
```

**Review by amesgen — APPROVED** [review]

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs (hunk 1 above)
Comment (jasagredo) [reply ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs]: Done

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs [hunk 2]
```diff
@@ -451,15 +452,26 @@ forkBlockForging IS{..} blockForging =
...
+            -- will remove all the transactions in that block from the mempool
+            -- as a defensive programming measure. Otherwise we'll run the risk
+            -- of forging the same invalid block again. This means that we'll
+            -- throw away some good transactions in the process.
+            case reason of
+              ChainDB.ValidationError (ExtValidationErrorLedger _) ->
+                whenJust
+                  (NE.nonEmpty (map (txId . txForgetValidated) txs))
+                  (lift . removeTxs mempool)
+              _ -> pure ()
```
Comment (nfrisby) [inline ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs:474]:
```suggestion
              ChainDB.InFutureExceedsClockSkew{} -> pure ()
              ChainDB.ValidationError err -> case err of
                ExtValidationErrorHeader{} -> pure ()
                ExtValidationErrorLedger{} ->
                  whenJust
                    (NE.nonEmpty (map (txId . txForgetValidated) txs))
                    (lift . removeTxs mempool)
```

**Review by nfrisby — APPROVED** [review]
I Approved, but I also made a suggestion that I feel somewhat strongly about.

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs (hunk 2 above)
Comment (jasagredo) [reply ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs]: Done

**Pull request merged**